### ISSUE TITLE
variableCasing_bear

### ISSUE THREAD

**Issue by hypothesist** [coala-analyzer/coala-bears#145]
Would be cool to have a bear that automatically applies camelCasing or snake_casing depending on what the user decides is the norm.

Would apply to classes, functions, variables. Renaming files would be nice too.

**Comment by AbdealiJK** [comment]
Is this for a specific language ? Or generic for all languages ?

**Comment by hypothesist** [comment]
That's the thing: it should be language independent when you look at it from the outside, but technically it's language dependent: each language has its own syntax, so there's no universal way to know which is a variable and which is a keyword.

**Comment by AbdealiJK** [comment]
You may want to check out the coalang stuff.
It aims to define a structure for languages to find out keywords and so on.
It seems to be the first step to do this. Using that you'd be able to
figure out whether a name is a class name, variable name, etc and then
applying the styling becomes simpler

**Comment by Makman2** [comment]
yeah though a universal AST would be the best choice (but we don't have one, coAST has still not advanced). Especially if you don't have kewords that signal a new variable like `var` in javascript or `Dim` in VB. In C you just have `type variable_name`, that's a bit more difficult^^
@AbdealiJK btw our coalang stuff isn't so advanced^^ there's some minor stuff^^

**Comment by hypothesist** [comment]
@AbdealiJK as @Makman2 pointed out and from what I saw in the code, coalang is still in its early stage. So I think I'll start with developing that (as that way the features useful for variable casing bear can be used in other bears too) and it should make this easy.

**Comment by sils1297** [comment]
that'd be awesome :)

**Comment by AbdealiJK** [comment]
Awesome :)

**Comment by hypothesist** [comment]
I have a question: isn't it mandatory for us to ask the user if he wants to change this variable or not for each violation? For instance say he uses Gtk+. It uses a lot of underscores. But if the user wants camelCase for his variables (and says so in the coafile), we can't change the library functions' names.

So my solution is: for each new variable (this is important: it doesn't make sense to ask the user for each occurrence of a variable; we'll ask just the first time) that violates their global rule, we'll ask them if they want to change it. If they say yes, we'll change it everywhere, else, never change that variable anywhere.

**Comment by abhsag24** [comment]
@hypothesist that'd be pretty  cool IMO :)

**Comment by AbdealiJK** [comment]
@hypothesist It makes sense to add exceptions into the coafile as a setting rather than asking them ...
Its change-able easily and persists across multiple runs (You can't expect them to do it every time they run coala)

**Comment by hypothesist** [comment]
Hmm, yeah. I'll also add globs there so that they can ignore multiple names at once (something like `g_*` would prevent flagging anything related to Gtk+ automatically)

**Comment by hypothesist** [comment]
Another problem: consider the file

```
int main() {
    int abc_def = 1
    int hij_klm = abc_def + 1
}
```
The problem here is that the bear would apply `abc_def -> abcDef` properly on both lines, but the `hij_klm` won't change unless you run coala again due to diff conflicts.

I have a solution that will finish the whole file in one run, but that will need the user's permission for each line where there is a violation (and if the user by mistake chooses to not apply the patch for a particular line, their program will not even compile due to undefined variables). This is extremely dangerous.

So I think the only solution to variable casing bear that I can think of (which is unlike most bears because this acts on multiple lines) is to run it multiple times: which if you do, it will work perfectly (it will prompt the user exactly once per variable, ...).

What do you think?

**Comment by hypothesist** [comment]
Also, when this is the case, what warning message should be displayed? Does `There are conflicting name changes. Run coala multiple times to automatically resolve this` sound okay?

**Comment by sils1297** [comment]
Ah I get your problem. You shouldn't deal with this, actually it'd be good if @fneu could take care of this in coala, dealing with changes in the same line in the Diff class generically.

**Comment by hypothesist** [comment]
Cool, but as of now, I'll push variable casing bear? I think it's complete :) Once @fneu solves that it's just a line or two of change.

**Comment by sils1297** [comment]
@hypothesist you should just push out a patch for each variable name change where the base file should be the original one! Always the original. coala will take care of the rests and your bear will later not need any changing, coala will just calculate offsets within the line if it changed.

**Comment by hypothesist** [comment]
@sils1297 I know, I just meant the warning message (and the code accompanying it to determine if we should give a warning). We should let the user know that some variables have not been changed. Once @fneu finishes it, we'll just remove that (because at that time it will work seamlessly).

**Comment by sils1297** [comment]
IMO you shouldn't give a warning at all.

**Comment by Adrianzatreanu** [comment]
@adtac !

**Comment by adtac** [comment]
Just waiting for the two year anniversary :P 

Oh man, does someone else want to take this over? I don't even remember half the stuff I've written. It's fairly polished, just a few more review comments (1 or 2 review iterations at max). Someone with more coala-bears experience?

**Comment by meetmangukiya** [comment]
no PR reference on the issue though

**Comment by adtac** [comment]
https://github.com/coala/coala-bears/pull/484 literally the oldest open PR lol :D
